### PR TITLE
Add declare rust

### LINKED ISSUE #1888: Implement `declare` cheatcode for rust test runner
We should implement `declare` cheatcode as a proving ground for other cheatcodes. It should declare cairo 1 contracts and use starknet in rust. It can be split into smaller tasks.

We probably should use https://github.com/lambdaclass/starknet_in_rust/blob/f4b80ed222381aea6e3f5a81047fa1c829d2fe97/crates/starknet-rs-py/src/starknet_state.rs#L35 to declare the contract.

We should build on 0.11 branch of starknet in rust.

We should use custon HintProcessor to handle cheatcode hints https://github.com/software-mansion-labs/cairo/blob/4094d447c77990142361f14b057f2e477a4d3130/crates/cairo-lang-runner/src/casm_run/mod.rs#L1315

### PR DESCRIPTION
**Pull request by karol-bisztyga**
https://github.com/software-mansion/protostar/issues/1888

cairo pr https://github.com/software-mansion-labs/cairo/pull/78

blockifier PR https://github.com/starkware-libs/blockifier/pull/540

### CODE REVIEW

File: protostar-rust/src/lib.rs [hunk 1]
```diff
@@ -42,9 +48,20 @@ fn internal_run_tests(input_path: &str) -> anyhow::Result<()> {
...
+    let path = paths[0].clone();
+    let key_converted: String = key.extract().unwrap();
+    contracts_paths.insert(key_converted, path);
+  }
+  let test_config = ProtostarTestConfig{ contracts_paths };
+    internal_run_tests(&input_path.to_owned(), test_config)
+        .map_err(|e| PyErr::new::<PyRuntimeError, _>(format!("{:?}", e)))?;
 
     Ok(())
 }
```
Comment (cptartur) [inline protostar-rust/src/lib.rs:67]: I'm not sure about extensive use of `unwrap` in module like that, but we can resolve that in Rust CLI.

File: protostar/commands/cairo1_commands/test_rust_command.py [hunk 2]
```diff
@@ -33,8 +40,18 @@ def arguments(self):
         ]
 
     async def run(self, args: Namespace):
+        contract_paths: dict[str, list[str]] = {}
```
Comment (cptartur) [inline protostar/commands/cairo1_commands/test_rust_command.py:43]:
```suggestion
        contract_paths: Dict[str, List[str]] = {}
```

I think it's better to use `Dict` and `List` from typing package here.

File: protostar/commands/cairo1_commands/test_rust_command.py [hunk 3]
```diff
@@ -33,8 +40,18 @@ def arguments(self):
...
+            contract_name
+        ) in self._contract_path_resolver.configuration_file.get_contract_names():
+            contract_paths[contract_name] = [
+                str(path)
+                for path in self._contract_path_resolver.configuration_file.get_contract_source_paths(
+                    contract_name
+                )
+            ]
         if args.path is None:
             raise ProtostarException("No tests provided")
```
Comment (cptartur) [inline protostar/commands/cairo1_commands/test_rust_command.py:54]: Is this assert needed? I think `path` should be made a required argument.

File: protostar/commands/cairo1_commands/test_rust_command.py [hunk 4]
```diff
@@ -33,8 +40,18 @@ def arguments(self):
         ]
 
     async def run(self, args: Namespace):
+        contract_paths: dict[str, list[str]] = {}
+        for (
+            contract_name
+        ) in self._contract_path_resolver.configuration_file.get_contract_names():
+            contract_paths[contract_name] = [
+                str(path)
+                for path in self._contract_path_resolver.configuration_file.get_contract_source_paths(
```
Comment (cptartur) [inline protostar/commands/cairo1_commands/test_rust_command.py:49]: I'm not sure about using `configuration_file` from ContractPathResolver like that. I'd be better to directly inject a correct `ConfigurationFile` as a `TestRustCommand` dependency and use that instead.

File: pyrightconfig.json [hunk 5]
```diff
@@ -0,0 +1,3 @@
+{
+  "ignore": ["./starknet_in_rust/", "./blockifier/"]
+}
```
Comment (cptartur) [inline pyrightconfig.json:3]: Let's add a newline here.

File: tests/e2e/cairo1/test_cairo1_testing_rust.py [hunk 6]
```diff
@@ -44,3 +44,17 @@ def test_failing_tests(protostar: ProtostarFixture, copy_fixture: CopyFixture):
         "test_panic_multiple_values: Panic [1870930782904301745253, 482670963043, 31066316372818838395891839589]"
         in result
     )
+
+
+def test_declare(protostar: ProtostarFixture, copy_fixture: CopyFixture):
```
Comment (cptartur) [inline tests/e2e/cairo1/test_cairo1_testing_rust.py:49]: Naming this `test_declare` is confusing considering we have a `protostar declare` command.

File: tests/e2e/fixtures/cairo1/declare_test.cairo [hunk 7]
```diff
@@ -0,0 +1,10 @@
+use result::ResultTrait;
+use protostar_print::PrintTrait;
+
+#[test]
+fn test_declare_simple() {
+    let class_hash = declare('main').unwrap();
+    assert(class_hash == 273, 'proper class hash');
+
+    assert(1 == 1, 'simple check');
+}
```
Comment (cptartur) [inline tests/e2e/fixtures/cairo1/declare_test.cairo:10]: Newline

File: tests/e2e/cairo1/test_cairo1_testing_rust.py (hunk 6 above)
Comment (cptartur) [reply tests/e2e/cairo1/test_cairo1_testing_rust.py:49]: I think `test_declare_cheatcode` or something like that would be better.

File: .circleci/config.yml
Comment (cptartur) [inline .circleci/config.yml:1]: I didn't do a very thorough review of this file. I'm assuming most of the code was based on already existing stuff so it should work.

File: pyrightconfig.json (hunk 5 above)
Comment (karol-bisztyga) [reply pyrightconfig.json:3]: I added this file in [kb/1906-add-test-rust-command](https://github.com/software-mansion/protostar/tree/kb/1906-add-test-rust-command) with a new line

File: protostar/commands/cairo1_commands/test_rust_command.py (hunk 2 above)
Comment (cptartur) [reply protostar/commands/cairo1_commands/test_rust_command.py:43]: I don't think this got resolved

**Review by cptartur — APPROVED** [review]

**Review by piotmag769 — APPROVED** [review]

**Comment by karol-bisztyga** [comment]
Closing, #2059

**Pull request closed without merging**